checkpoint_engine/ps.py

```diff
@@ -576,7 +581,80 @@ def _gen_h2d_buckets(
         assert buckets[-1][1].size > 0, (
             f"buckets[-1][1].size {buckets[-1][1].size} should be greater than 0"
         )
-    return buckets
+    actual_local_topo = (
+        {k: v & set(ranks) for k, v in local_topo.items() if v & set(ranks)}
+        if ranks
+        else local_topo
+    )
+    # if ranks is empty, assign the owner_rank as receiver_rank, this is used for colocate architecture
+    if not ranks:
+        return [(owner_rank, owner_rank, bucket) for owner_rank, bucket in buckets]
+    else:
+        return _assign_receiver_ranks(buckets, actual_local_topo, remote_topo)
+
+
+if TYPE_CHECKING:
```

Comment:
This TYPE_CHECKING is existed above, this definition of `T` should move into that block